[#24971] Add a retry policy for JmsIO #24971

Amraneze opened the pull request:
Add a retry policy to publish messages in case of connection closed

Fixes #24971

Co-Authored-By: Amrane Ait Zeouay <122456352+amranezeRenault@users.noreply.github.com>

When trying to publish messages using `JmsIO` with a closed/failed connection, the messages are not published and lost. We will need to get the failed messages with the function `getFailedMessages` and try again and again multiple times. 

````java
var failedPublishedMessages = messages.apply(getWriter(sinkOptions)).getFailedMessages();
var refailedPublishedMessages = failedPublishedMessages.apply(REPUBLISH_FAILED_MESSAGE_NAME, getWriter(sinkOptions)).getFailedMessages();
....
return PDone.in(cdpMessages.getPipeline());
````

To not have multiple steps for retrying the failed publication, we added a `Retry Policy` which consist on retrying the publication n times (the number is given by the user) until the timeout (given by the user too) is finished. Also, we added a predicate function, to give the user the option to handle when the connection should be reconnected or not in case of a failed connection.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

Abacn commented:
just fyi beam has backoff utility classes for retry: 
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/util/BackOffUtils.java
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java

and is quite used in IOs and can be referred as retry implementations.

Amraneze commented:
> 

The issue is that JmsIO doesn't retry at all when publishing a message, any message that is failed to publish is pushed to ouput as you can see in [JmsIO code](https://github.com/apache/beam/blob/master/sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java#L940):

````java
@ProcessElement
public void processElement(ProcessContext ctx) {
  Destination destinationToSendTo = destination;
  try {
    Message message = spec.getValueMapper().apply(ctx.element(), session);
    if (spec.getTopicNameMapper() != null) {
      destinationToSendTo =
          session.createTopic(spec.getTopicNameMapper().apply(ctx.element()));
    }
    producer.send(destinationToSendTo, message);
  } catch (Exception ex) {
    LOG.error("Error sending message on topic {}", destinationToSendTo);
    ctx.output(failedMessageTag, ctx.element());
  }
}
````
If there are any exceptions, they will be catched and the message is pushed to failed message tag's output. Which means we will need to get the failed messages and retry by ourselves N times like this:

````java
public PDone expand(PCollection<Message> messages) {
    // Retry 3 times when the session is closed
    messages.apply(getJmsWriter(sinkOptions))
          .getFailedMessages()
         .apply(REPUBLISH_FAILED_MESSAGE_NAME, getJmsWriter(sinkOptions))
           .getFailedMessages()
         .apply(REPUBLISH_FAILED_MESSAGE_NAME, getJmsWriter(sinkOptions))
           .getFailedMessages()
         .apply(REPUBLISH_FAILED_MESSAGE_NAME, getJmsWriter(sinkOptions));

    return PDone.in(messages.getPipeline());
  }

  private static JmsIO.Write<Message> getJmsWriter(SinkOptions sinkOptions) {
    return JmsIO.<Message>write()
        .withConnectionFactory(SinkOptions.getConnectionFactory())
        .withValueMapper(getValueMapper())
        .withTopicNameMapper(getTopicNameMapper());
  }
````
When the session is closed and the first PTransform that is publishing the message is displaying error logs `Error sending message on topic`, it will not reconnect and just send all messages to the second step as you can see in the screenshot

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/28459763/213120382-941c1769-805d-4685-aff0-94a508b94f3d.png">

>Note: The graph of first PTransform is the same as the second PTransform after the session is closed, which means the first step doesn't publish at all and it just send all messages to the second step.

Abacn commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIOProducer.java:
```diff
@@ -0,0 +1,222 @@
+    public void setup() {
+      retryPublicationBackoff =
+          FluentBackoff.DEFAULT
+              .withMaxRetries(0)
```
@Amraneze thanks for explanation. I get where the issue is. What I mean previously ""by default" was that, the sink should retry by default. Here I see if retrypolicy is not setup, there is still no retry (current behavior)

I am suggesting follow the same pattern used elsewhere (e.g. https://github.com/apache/beam/blob/aba175259944caa75a7673b9e0ad430d8e5c68fe/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java#L2459-L2465)

Specifically, name the retry policy as RetryConfiguration that is an internal class of JmsIO, and set a default retry configuration if user does not provide it. I see there is a substantial refactoring of the code in this PR which may not be necessary (producer can still be in JmsIO class) finally thanks that the unit test has been implemented.

Amraneze replied:
Ok, I will work on it.

Amraneze replied:
@Abacn Can you recheck if everything is good ?

Abacn commented on sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/FakeConnection.java:
```diff
@@ -0,0 +1,100 @@
+import org.apache.activemq.ConnectionFailedException;
+import org.mockito.Mockito;
+
+public class FakeConnection implements Connection, Serializable {
```
(minor) though been test class, nice to have a javadoc saying what this mock class achieves.

Abacn reviewed (changes requested):
Thanks for the change! I see what has caused the issue, and commented above that there are changes still needed in order to fix it. Please let me know if it is not clear to you or if there is any questions.

Amraneze commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -932,36 +1067,78 @@ public void setup() throws Exception {
         }
       }
 
+      private void restartJmsConnection() throws JMSException {
```
When the connection is lost or closed, the JMS client calls the `ExceptionListener` if it exists. I added that listener 

````java
this.connection.setExceptionListener(
  exception -> {
    if (spec.getConnectionFailedPredicate() != null && spec.getConnectionFailedPredicate().apply(exception)) {
      LOG.error("Jms connection encountered the following exception:", exception);
      connectionErrors.inc();
      try {
        restartJmsConnection();
      } catch (JMSException e) {
        LOG.error("An error occurred while reconnecting:", e);
      }
    }
});
````
Should I just re-throw the exception so that the DoFn will call the `FinishBundle` method ? Something like this 

````java
this.connection.setExceptionListener(exception -> {
    LOG.error("Jms connection encountered the following exception:", exception);
    connectionErrors.inc();
    throw new JmsIOException("The connection was closed", exception);
});
````
Or what do you think we should do?

Abacn replied:
Yes, should re-throw the error then let the runner retry the bundle, otherwise the bundle will still success but data not got written

Amraneze commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -932,36 +1067,78 @@ public void setup() throws Exception {
-        connection = null;
+      public void teardown() throws JMSException {
+        if (producer != null) {
+          producer.close();
```
With the current implementation of the producer, we don't have a way to flush the messages. Should we create an accumulative list of messages and when tearing down the DoFn we can flush those messages, what do you think ?

>Note: When the connection is closed, the producer is no longer available, so we will need to recreate a connection, a session and the producer. Which means that during Finish bundle DoFn's lifecycle, we will need to recreate them to flush the messages. That's why I called the two functions `teardown` & `start` to make sure the producer is recreated in case of an exception even though it changes the mutation of the objects.

<img width="466" alt="image" src="https://user-images.githubusercontent.com/28459763/214697936-91f1461a-9efc-44d1-8347-fcf62ac46f91.png">

Abacn replied:
`@TearDown` annotated method should not handle any data related operation, because there is no guarantee that `@TearDown` method will be called, and a fail does not trigger runner retry so data loss is highly likely. If there is no way to flush a client without closing it, then yes I think we need to restart it everytime.

Amraneze replied:
There is still one thing that perplexed me, if a connection is closed and an exception is thrown, the DoFn will call FinishBundle. Which case is the correct one ?
- The messages of that DoFn are lost and the retry configuration won't be helpful, because it restarts the whole thread.
- The DoFn restart the cycle but keeps in state the old messages while retying the publish in the `processElement` in a different thread.

Abacn replied:
If a connection is closed and an exception is thrown in processElement, DoFn will not call finishbundle imediately, it will retry the whole bundle (restarting from startBundle)

Abacn commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -897,71 +944,154 @@ private boolean isExclusiveTopicQueue() {
+          session.close();
+          session = null;
+        }
+        if (connection != null) {
```
I am thinking about the possible overhead of close and re-establish all of producer, session, and connection in each bundle. Does close the producer alone suffices flushing the message (committing them to server)? If so we can just close the producer, and close session and connections in tearDown.

Abacn reviewed (commented):
Thanks for the change. got a couple of comments should be easy to resolve. Thanks for your patience.

Amraneze commented:
Hello @Abacn, Sorry I was busy with work. I redesigned the flow of `JmsIOWriter`. I tried to create a cache and flush all the messages in `finishBundle` but we will need to handle the exception in `processElement` otherwise the DoFn will tear down without calling `finishBundle`. I added a new step which will be used only for processing failed published messages and also I added a condition whenever the producer needs to be recreated it. Let me know what do you think and if is there any tests to add ?
>PS: Should I add a benchmark test ?

Abacn commented:
Sorry for the delay. Will take another look tomorrow.

Abacn commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -932,36 +1036,158 @@ public void setup() throws Exception {
+      }
+    }
+
+    private static class JmsIOProducerFn<T> extends DoFn<T, T> {
```
If I understand correctly, if should just work using JmsIOProduceRetryFn alone, why need another `JmsIOProducerFn` before it?

I personally think two step retry-write-on-failed is not a good pattern because it is unnecessarily complicated and more important it is hard to test (though I see it tested in unit test). It also requires to maintain two classes doing the same thing and keep them consistent for any future addition which is errorprone.

Amraneze replied:
I separated them because I'm not sure if the Backoff will hold the DoFn thread and I don't know if it will do that for the whole bundle (Which means the whole bundle will be in wait until the failed message is published). Just to not have any performances issues doing that. Do you think it would be better to keep in one DoFn ?

>PS: I tried to have a sort of cache of failed messages and flush them in Finishbundle cycle, but when throwing an exception in ProcessElement, the worker will call TearDown without FinishBundle.

Abacn replied:
For messages go to JmsIOProduceRetryFn and have no error, if does not trigger backoff and do not hold DoFn thread.

Currently, if error happens, JmsIOProducerFn send failedMessage to JmsIOProduceRetryFn, if failed again it still triggers backoff. In runner JmsIOProducerFn and JmsIOProduceRetryFn will be fused and run on same thread. So if this is a concern it exists currently also

Amraneze replied:
Ok, so what's the best approach to keep the performance as it was ? should I just merge them into one ?

Abacn replied:
For now I would simply use `JmsIOProduceRetryFn`. For succeeded send, the while (true) loop just breaks before checking backOff, it should has same performance as `JmsIOProducerFn`.

There are of course way of improvement. Currently I understand the `.send()` call is synchroneous (otherwise backoff won't work). Switch to cached, asynchronous API call would be more performant. An example in Beam's code base is KafkaWriter: https://github.com/apache/beam/blob/master/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriter.java

Amraneze replied:
@Abacn Maybe we can for now merge the current solution, I will try to find time to make it better. I will keep in mind the example of kafka writer. I pushed a new commit with your recommendations.

Abacn commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -897,32 +944,89 @@ private boolean isExclusiveTopicQueue() {
+    }
 
-    private static class WriterFn<EventT> extends DoFn<EventT, EventT> {
+    private static class JmsConnection<T> implements Serializable {
```
thanks, a helper class really reduced the complexity of DoFns.

Abacn commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -932,36 +1036,158 @@ public void setup() throws Exception {
+          } catch (JMSException | JmsIOException exception) {
+            if (!BackOffUtils.next(sleeper, backoff)) {
+              LOG.error("Error sending message", exception);
+              context.output(this.failedMessagesTags, input);
```
already have LOG.error and output here.

Amraneze replied:
Yeah, I was thinking about it, I will remove it we know that the message is sent to the output, I wanted just to log that the message was failed.

Abacn commented on sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:
```diff
@@ -932,36 +1036,158 @@ public void setup() throws Exception {
+        try {
+          publishMessage(input, context);
+        } catch (IOException | InterruptedException exception) {
+          LOG.error("Error while publishing the message", exception);
```
Outputed twice and double logging for failed publishMessage (see below)

publishMessage may only publish messege (no need ProcessContext parameter their) and throw exception when retry count used up. Then deal with the failure in processElement here.

Amraneze replied:
Ok, I will do that

Amraneze replied:
Something like this

````java
@ProcessElement
public void processElement(@Element T input, ProcessContext context) {
  try {
    publishMessage(input);
  } catch (JMSException | JmsIOException | IOException | InterruptedException exception) {
    LOG.error("Error while publishing the message", exception);
    context.output(this.failedMessagesTags, input);
    if (exception instanceof InterruptedException) {
      Thread.currentThread().interrupt();
    }
  }
}

private void publishMessage(T input)
    throws JMSException, JmsIOException, IOException, InterruptedException {
  Sleeper sleeper = Sleeper.DEFAULT;
  BackOff backoff = checkStateNotNull(retryBackOff).backoff();
  while (true) {
    try {
      this.jmsConnection.publishMessage(input);
      break;
    } catch (JMSException | JmsIOException exception) {
      if (!BackOffUtils.next(sleeper, backoff)) {
        throw exception;
      } else {
        publicationRetries.inc();
      }
    }
  }
}
````

Abacn replied:
yes that looks good

Abacn reviewed (changes requested):
Thanks, it's getting close. Found a bug and a design suggestion here.

Abacn reviewed (approved):
Thanks. This LGTM.

nit: the added back-off unit tests take a while to complete. In the case of test may use a mocked sleeper

Amraneze commented:
> Thanks. This LGTM.
> 
> nit: the added back-off unit tests take a while to complete. In the case of test may use a mocked sleeper

I will improve the test with the new API calls.

Amraneze commented:
@Abacn can you please merge the PR, I don't have access to merge it :( ?
